File: codes/scala/chapter_backtracking/permutations_ii.scala
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

/**
 * File: permutations_ii.scala
 * Author: WangBlue (wangjiping596@gmail.com)
 * Create: 2024-08-16 14:20
 */

object permutations_ii {
  /* 回溯算法：全排列 I */
  def backtrack(state: List[Int], choices: Array[Int],
                selected: Array[Boolean],res: ListBuffer[List[Int]]): Unit = {
    // 当状态长度等于元素数量时，记录解
    if(state.size == choices.length) {
      res += state
      return
    }
    // 遍历所有选择
    val duplicated = new mutable.HashSet[Int]()
    for(i <- choices.indices) {
      val choice = choices(i)
      // 剪枝：不允许重复选择元素
      if(!selected(i) && !duplicated.contains(choice)) {
        // 尝试：做出选择，更新状态
        selected(i) = true
        duplicated += choice
        // 进行下一轮选择
        backtrack(state :+ choice, choices, selected, res)
        // 回退：撤销选择，恢复到之前的状态
        selected(i) = false
      }
    }
  }

  /* 全排列 I */
  def permutationsII(nums: Array[Int]): List[List[Int]] = {
    val res = ListBuffer[List[Int]]()
    backtrack(List(), nums, Array.fill(nums.length)(false), res)
    res.toList
  }

  def main(args: Array[String]): Unit = {
    // 定义输入数组
    val nums: Array[Int] = Array(1, 2, 2)

    // 调用全排列方法
    val res: List[List[Int]] = permutationsII(nums)

    // 打印结果
    println(s"输入数组 nums = ${nums.mkString(", ")}")
    println(s"所有排列 res = ${res.map(_.mkString("[", ", ", "]")).mkString(", ")}")
  }
}

File: ru/codes/c/chapter_backtracking/permutations_ii.c
/**
 * File: permutations_ii.c
 * Created Time: 2023-10-17
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.h"

// Предположить, что существует не более 1000 перестановок, а максимальный элемент равен 1000
#define MAX_SIZE 1000

/* Алгоритм бэктрекинга: все перестановки II */
void backtrack(int *state, int stateSize, int *choices, int choicesSize, bool *selected, int **res, int *resSize) {
    // Когда длина состояния равна числу элементов, записать решение
    if (stateSize == choicesSize) {
        res[*resSize] = (int *)malloc(choicesSize * sizeof(int));
        for (int i = 0; i < choicesSize; i++) {
            res[*resSize][i] = state[i];
        }
        (*resSize)++;
        return;
    }
    // Перебор всех вариантов выбора
    bool duplicated[MAX_SIZE] = {false};
    for (int i = 0; i < choicesSize; i++) {
        int choice = choices[i];
        // Отсечение: нельзя выбирать один и тот же элемент повторно и нельзя повторно выбирать равные элементы
        if (!selected[i] && !duplicated[choice]) {
            // Попытка: сделать выбор и обновить состояние
            duplicated[choice] = true; // Записать значения уже выбранных элементов
            selected[i] = true;
            state[stateSize] = choice;
            // Перейти к следующему выбору
            backtrack(state, stateSize + 1, choices, choicesSize, selected, res, resSize);
            // Откат: отменить выбор и восстановить предыдущее состояние
            selected[i] = false;
        }
    }
}

/* Все перестановки II */
int **permutationsII(int *nums, int numsSize, int *returnSize) {
    int *state = (int *)malloc(numsSize * sizeof(int));
    bool *selected = (bool *)malloc(numsSize * sizeof(bool));
    for (int i = 0; i < numsSize; i++) {
        selected[i] = false;
    }
    int **res = (int **)malloc(MAX_SIZE * sizeof(int *));
    *returnSize = 0;

    backtrack(state, 0, nums, numsSize, selected, res, returnSize);

    free(state);
    free(selected);

    return res;
}

/* Driver Code */
int main() {
    int nums[] = {1, 1, 2};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int returnSize;

    int **res = permutationsII(nums, numsSize, &returnSize);

    printf("Входной массив nums = ");
    printArray(nums, numsSize);
    printf("\nВсе перестановки res = \n");
    for (int i = 0; i < returnSize; i++) {
        printArray(res[i], numsSize);
    }

    // Освободить память
    for (int i = 0; i < returnSize; i++) {
        free(res[i]);
    }
    free(res);

    return 0;
}
